File: 04- April/25- Smallest Number in Infinite Set/25- Smallest Number in Infinite Set (Noura Algohary).cpp
//Author: Noura Algohary

class SmallestInfiniteSet {
public:
    // set is used to store UNIQUE values in ascending order by default
    set<int>st;
    SmallestInfiniteSet() {
        st.clear();

        for(int i = 1; i<=1000;i++)
            st.insert(i);
    }
    
    int popSmallest() {
        // smallest element is at the beginning of the set 
        int removed = *st.begin();
        st.erase(st.begin());

        return removed;
    }
    
    void addBack(int num) {
        st.insert(num);
    }
};

/**
 * Your SmallestInfiniteSet object will be instantiated and called as such:
 * SmallestInfiniteSet* obj = new SmallestInfiniteSet();
 * int param_1 = obj->popSmallest();
 * obj->addBack(num);
 */

File: 04- April/25- Smallest Number in Infinite Set/25- Smallest Number in Infinite Set (Noura Algohary).py
# Author: Noura Algohary
class SmallestInfiniteSet:
    # use list as a set 
    st = []

    def __init__(self):
        self.st.clear()

        for i in range(1, 1001):
            self.st.append(i)
        

    def popSmallest(self) -> int:
        # sort to return the first element(smallest element)
        self.st.sort()

        removed = self.st.pop(0)

        return removed

    def addBack(self, num: int) -> None:
        # make sure the number is not presented in the set
        if num not in self.st:
            self.st.insert(0, num)
        


# Your SmallestInfiniteSet object will be instantiated and called as such:
# obj = SmallestInfiniteSet()
# param_1 = obj.popSmallest()
# obj.addBack(num)